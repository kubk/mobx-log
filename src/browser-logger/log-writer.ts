export type LogWriter = {
  write: (...messages: unknown[]) => void;
};

export class DefaultLogWriter implements LogWriter {
  write(...messages: unknown[]) {
    console.log(...messages);
  }
}

export class CollectingLogWriter implements LogWriter {
  history: string[] = [];

  write(...messages: unknown[]) {
    const strings = messages.map((message): string =>
      (message as any).toString()
    );
    this.history.push(strings.join(' '));
  }

  clear() {
    this.history = [];
  }
}
