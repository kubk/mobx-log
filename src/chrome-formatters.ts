import {
  isObservableArray,
  isObservableMap,
  isObservableObject,
  isObservableSet,
  ObservableMap,
  ObservableSet,
  toJS,
} from 'mobx';

type Css = { style: string };
type Chunk = [string, Css, string];
type Header = [string, Chunk, [string, string]];
type Body = [string, Css, ...any[]];

export type ChromeFormatter<T> = {
  header: (argument: T) => Header | null;
  hasBody: (argument: T) => boolean;
  body: (argument: T) => Body;
};

const styles = {
  mobxName: { style: 'color: rgb(232,98,0)' },
  nullName: { style: 'color: #777' },
  list: {
    style:
      'list-style-type: none; padding: 0; margin: 0 0 0 12px; font-style: normal',
  },
};

const reference = (object: unknown) => {
  if (typeof object === 'undefined' || typeof object === null) {
    return ['span', styles.nullName, typeof object];
  }
  return ['object', { object }];
};

const renderIterableHeader = (
  iterable:
    | ArrayLike<unknown>
    | ObservableMap<unknown, unknown>
    | ObservableSet<unknown>,
  name: string
): Header => {
  const count =
    isObservableMap(iterable) || isObservableSet(iterable)
      ? iterable.size
      : iterable.length;

  return ['span', ['span', styles.mobxName, name], ['span', `[${count}]`]];
};

const renderIterableBody = (
  collection:
    | ArrayLike<unknown>
    | ObservableMap<unknown, unknown>
    | ObservableSet<unknown>
    | {},
  mapper: (val: [string, unknown]) => unknown
): Body => {
  if (isObservableMap(collection) || isObservableSet(collection)) {
    // @ts-ignore
    const children = Array.from(toJS(collection)).map(mapper);
    return ['ol', styles.list, ...children];
  }

  const children = Object.entries(toJS(collection)).map(mapper);
  return ['ol', styles.list, ...children];
};

export class ObjectFormatter implements ChromeFormatter<{}> {
  header(argument: {}): Header | null {
    return isObservableObject(argument)
      ? renderIterableHeader(Object.keys(argument), 'Object')
      : null;
  }

  hasBody(argument: {}): boolean {
    return Object.keys(argument).length > 0;
  }

  body(argument: {}) {
    return renderIterableBody(argument, ([key, value]) => [
      'li',
      {},
      reference(key),
      ': ',
      reference(value),
    ]);
  }
}

export class ArrayFormatter implements ChromeFormatter<ArrayLike<unknown>> {
  header(argument: ArrayLike<unknown>): Header | null {
    return isObservableArray(argument)
      ? renderIterableHeader(argument, 'Array')
      : null;
  }

  hasBody(argument: ArrayLike<unknown>): boolean {
    return argument.length > 0;
  }

  body(argument: ArrayLike<unknown>): Body {
    return renderIterableBody(argument, ([_, value]) => [
      'li',
      reference(value),
    ]);
  }
}

export class MapFormatter implements ChromeFormatter<Map<unknown, unknown>> {
  header(argument: Map<unknown, unknown>): Header | null {
    return isObservableMap(argument)
      ? renderIterableHeader(argument, 'Map')
      : null;
  }

  hasBody(argument: Map<unknown, unknown>): boolean {
    return argument.size > 0;
  }

  body(argument: Map<unknown, unknown>): Body {
    return renderIterableBody(argument, ([key, value]) => [
      'li',
      {},
      reference(key),
      ': ',
      reference(value),
    ]);
  }
}

export class SetFormatter implements ChromeFormatter<Set<unknown>> {
  header(argument: Set<unknown>): Header | null {
    return isObservableSet(argument)
      ? renderIterableHeader(argument, 'Set')
      : null;
  }

  hasBody(argument: Set<unknown>): boolean {
    return argument.size > 0;
  }

  body(argument: Set<unknown>): Body {
    return renderIterableBody(argument, (value) => ['li', reference(value)]);
  }
}
