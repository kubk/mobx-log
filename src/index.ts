export { makeLoggable } from './make-loggable';
export { DefaultLogger } from './default-logger';
export type {
  ObservableEvent,
  ComputedEvent,
  ActionEvent,
  Logger,
} from './types';
export { DefaultLogWriter } from './log-writer';
export { configureMakeLoggable } from './config';
export { useLoggableLocalObservable } from './use-loggable-local-observable';
export { installMobxFormatters as applyFormatters } from './chrome-formatters';
