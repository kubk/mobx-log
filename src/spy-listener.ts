import { comparer, spy } from 'mobx';
import type { PureSpyEvent } from 'mobx/dist/core/spy';
import { Logger } from './types';
import { getStoreName, isStore } from './store';

export class SpyListener {
  private filtersByClass: string[] = [];

  constructor(private logger: Logger, private debug = false) {}

  listen() {
    spy(this.log);
  }

  private log = (event: PureSpyEvent) => {
    const { logger } = this;

    if (this.debug) {
      console.log(event);
    }

    if (event.type === 'update') {
      if (event.observableKind === 'computed') {
        const equals = comparer.default;
        if (!equals(event.oldValue, event.newValue)) {
          const computedFullName = this.parseDebugName(event.debugObjectName);
          const [storeName] = computedFullName.split('.');
          if (!this.filtersByClass.includes(storeName)) {
            return;
          }
          logger.logComputed({
            name: computedFullName,
            oldValue: event.oldValue,
            newValue: event.newValue,
          });
        }
      }
      if (event.observableKind === 'object') {
        const observableFullName = this.parseDebugName(event.debugObjectName);
        const [storeName] = observableFullName.split('.');
        if (!this.filtersByClass.includes(storeName)) {
          return;
        }
        // Mobx spy adds '[..]' to object name in case change happened inside array
        const isArray = observableFullName.endsWith('[..]');
        const name = `${
          isArray ? observableFullName : storeName
        }.${event.name.toString()}`;

        logger.logObservable({
          type: 'update',
          name,
          newValue: event.newValue,
          oldValue: event.oldValue,
        });
      }
    }

    if (event.type === 'action') {
      if (typeof event.object !== 'object') {
        return;
      }
      if (!isStore(event.object)) {
        return;
      }
      const storeName = getStoreName(event.object);
      if (!this.filtersByClass.includes(storeName)) {
        return;
      }

      const action =
        'loggableName' in event.object ? event.name.split('.')[1] : event.name;

      logger.logAction({
        name: `${storeName}.${action}`,
        arguments: event.arguments,
      });
      return;
    }

    if ('observableKind' in event) {
      if (event.observableKind === 'array') {
        const observableFullName = this.parseDebugName(event.debugObjectName);
        const [storeName, key] = observableFullName.split('.');
        if (!this.filtersByClass.includes(storeName)) {
          return;
        }
        logger.logObservable({
          type: event.observableKind,
          name: `${storeName}.${key.toString()}`,
          // @ts-expect-error
          removed: event.removed,
          // @ts-expect-error
          added: event.added,
        });
      }
      if (event.observableKind === 'map') {
        const observableFullName = this.parseDebugName(event.debugObjectName);
        const [storeName, key] = observableFullName.split('.');
        if (!this.filtersByClass.includes(storeName)) {
          return;
        }
        logger.logObservable({
          type: 'map',
          key: event.name,
          value: 'newValue' in event ? event.newValue : undefined,
          oldValue: 'oldValue' in event ? event.oldValue : undefined,
          name: `${storeName}.${key.toString()}`,
          mapType: event.type,
        });
      }
      if (event.observableKind === 'set') {
        const observableFullName = this.parseDebugName(event.debugObjectName);
        const [storeName, key] = observableFullName.split('.');
        if (!this.filtersByClass.includes(storeName)) {
          return;
        }
        const value =
          event.type === 'delete'
            ? event.oldValue
            : event.type === 'add'
            ? event.newValue
            : undefined;

        logger.logObservable({
          type: 'set',
          value,
          name: `${storeName}.${key.toString()}`,
          setType: event.type,
        });
      }
    }
  };

  private parseDebugName(name: string) {
    return name.replace(/@\d+/, '');
  }

  addFilterByStoreName(name: string) {
    this.filtersByClass.push(name);
  }
}
