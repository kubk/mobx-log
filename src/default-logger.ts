import { ActionEvent, ComputedEvent, Logger, ObservableEvent } from './types';
import { LogWriter } from './log-writer';

const colors = {
  red: '#FF6157',
  blue: '#28B7F6',
  orange: '#FFBC2D',
  green: '#2BC941',
};

export class DefaultLogger implements Logger {
  constructor(private logWriter: LogWriter) {}

  logObservable(event: ObservableEvent) {
    const info = [
      '%c[O]',
      `color:${colors.orange}`,
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }

  logAction(event: ActionEvent) {
    const formattedArguments =
      event.arguments.length > 0
        ? '%o '.repeat(event.arguments.length).slice(0, -1)
        : '';

    const info: unknown[] = [
      `%c[A]%c ${event.name}(${formattedArguments})`,
      `color:${colors.red}`,
      'color:inherit',
      ...event.arguments,
    ];

    this.logWriter.write(...info);
  }

  logComputed(event: ComputedEvent) {
    const info = [
      '%c[C]',
      `color:${colors.blue}`,
      event.name,
      event.oldValue,
      '->',
      event.newValue,
    ];

    this.logWriter.write(...info);
  }
}
