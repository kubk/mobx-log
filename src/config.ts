import { DefaultLogger } from './default-logger';
import { Logger } from './types';
import { DefaultLogWriter } from './log-writer';

export type Config = {
  logger: Logger;
  debug: boolean;
  storeConsoleAccess: boolean;
  filters: {
    computeds: boolean;
    actions: boolean;
    observables: boolean;
  }
};

export const config: Config = {
  logger: new DefaultLogger(new DefaultLogWriter()),
  debug: false,
  storeConsoleAccess: false,
  filters: {
    computeds: true,
    actions: true,
    observables: true,
  }
};

export const configureMakeLoggable = (options: Partial<Config>) => {
  if (options.logger !== undefined) {
    config.logger = options.logger;
  }
  if (options.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
  if (options.filters !== undefined) {
    config.filters = options.filters;
  }
};
