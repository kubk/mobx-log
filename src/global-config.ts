import { DefaultLogger } from './browser-logger/default-logger';
import { Logger } from './types';
import { DefaultLogWriter } from './browser-logger/log-writer';

export enum LoggerType {
  ReduxDevtools = 'redux-devtools',
  BrowserConsole = 'browser-console',
}

export type GlobalConfig = {
  logger: Logger;
  debug: boolean;
  type: LoggerType | null;
  storeConsoleAccess: boolean;
  filters: {
    computeds: boolean;
    actions: boolean;
    observables: boolean;
  };
};

export const config: GlobalConfig = {
  logger: new DefaultLogger(new DefaultLogWriter()),
  type: null,
  debug: false,
  storeConsoleAccess: false,
  filters: {
    computeds: true,
    actions: true,
    observables: true,
  },
};

type DevToolsOptions = Partial<
  Pick<GlobalConfig, 'debug' | 'storeConsoleAccess'>
>;

export const configureDevtools = (options?: DevToolsOptions) => {
  config.type = LoggerType.ReduxDevtools;
  if (options?.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options?.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
};

type LoggerOptions = Partial<
  Pick<GlobalConfig, 'logger' | 'debug' | 'storeConsoleAccess' | 'filters'>
>;

export const configureLogger = (options?: LoggerOptions) => {
  config.type = LoggerType.BrowserConsole;
  if (options?.logger !== undefined) {
    config.logger = options.logger;
  }
  if (options?.debug !== undefined) {
    config.debug = options.debug;
  }
  if (options?.storeConsoleAccess !== undefined) {
    config.storeConsoleAccess = options.storeConsoleAccess;
  }
  if (options?.filters !== undefined) {
    config.filters = options.filters;
  }
};
