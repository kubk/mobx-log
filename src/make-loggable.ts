import { SpyListener } from './spy-listener';
import { config } from './config';
import {
  installMobxFormatters,
  ChromeFormatter
} from './chrome-formatters';
import { getStoreName } from './store';
import { isObservable } from 'mobx';

declare global {
  interface Window {
    devtoolsFormatters: ChromeFormatter<any>[];
    store?: Record<string, object | Array<object>>;
  }
}

let spyListener: SpyListener;

const firstLetterLowerCase = (word: string) => {
  return word.charAt(0).toLowerCase() + word.slice(1);
};

export const makeLoggable = <T extends {}>(store: T): T => {
  if (process.env.NODE_ENV === 'production') {
    return store;
  }

  if (!isObservable(store)) {
    throw new Error(
      'mobx-log: store is not observable: ' +
        JSON.stringify(store, null, 2) +
        '. Make sure you called makeAutoObservable/makeObservable before calling makeLoggable'
    );
  }

  if (!spyListener) {
    spyListener = new SpyListener(config.logger, config.debug);
    spyListener.listen();

    installMobxFormatters()
  }

  const storeName = getStoreName(store);
  if (storeName === null) {
    return store;
  }

  if (config.storeConsoleAccess) {
    window.store = window.store || {};
    const storeKey = firstLetterLowerCase(storeName);
    window.store[storeKey] = store;
  }

  spyListener.addFilterByStoreName(storeName);

  return store;
};
