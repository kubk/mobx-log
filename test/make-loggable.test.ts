import { autorun, makeAutoObservable } from 'mobx';
import { DefaultLogger } from '../src';
import { configureMakeLoggable, makeLoggable } from '../src';
import { CollectingLogWriter } from '../src/log-writer';

const collectingWriter = new CollectingLogWriter();

configureMakeLoggable({
  logger: new DefaultLogger(collectingWriter),
});

class StoreWithComputed {
  value = 0;

  constructor() {
    makeAutoObservable(this);
    makeLoggable(this);
  }

  increment() {
    this.value++;
  }

  get isDividedBy3() {
    return this.value % 3 === 0;
  }
}

class StoreOnlyObservables {
  age?: number;
  name?: string;
  numbers: number[] = [];
  i = 0;

  constructor() {
    makeAutoObservable(this, {
      i: false,
    });
    makeLoggable(this);
  }

  init() {
    this.age = 24;
    this.name = 'Test';
  }

  add() {
    this.numbers.push(++this.i);
  }

  setMultiple() {
    this.numbers = [1, 2, 3, 4, 5, 6];
  }

  addAndRemoveSameAction() {
    this.numbers.push(++this.i);
    this.numbers.shift();
  }

  removeFirst() {
    this.numbers.shift();
  }

  removeLast() {
    this.numbers.pop();
  }

  spliceFirst() {
    this.numbers.splice(0, 1);
  }
}

class ParticipantStore {
  onlineUsers = new Map<number, boolean>();
  adminUsers = new Set<number>();

  constructor() {
    makeAutoObservable(this);
    makeLoggable(this);
  }

  onlineUserSet(id: number, value: boolean) {
    this.onlineUsers.set(id, value);
  }

  onlineUserDelete(id: number) {
    this.onlineUsers.delete(id);
  }

  adminUserAdd(value: number) {
    this.adminUsers.add(value);
  }

  adminUserDelete(value: number) {
    this.adminUsers.delete(value);
  }
}

class StoreWithoutLog {
  counter = 1;

  constructor() {
    makeAutoObservable(this);
  }

  increment() {
    this.counter++;
  }

  get isEven() {
    return this.counter % 2 === 0;
  }
}

type Todo = { id: number; text: string; isDone: boolean };

class TodoStore {
  todos: Todo[] = [];

  constructor() {
    makeAutoObservable(this);
    makeLoggable(this);
  }

  addTodo(id: number, text: string) {
    this.todos.push({
      id,
      text,
      isDone: false,
    });
  }

  markAsDone(id: number) {
    const todo = this.todos.find((todo) => todo.id === id);
    if (!todo) {
      throw new Error('Find error');
    }
    todo.isDone = true;
  }
}


export const createCounterStore = () => {
  return makeLoggable(
    makeAutoObservable({
      loggableName: 'counter',
      count: 0,
      increment() {
        this.count++;
      },
    })
  );
};

describe('makeLoggable', () => {
  beforeEach(() => {
    collectingWriter.clear();
  });

  it('logs', () => {
    const storeWithComputed = new StoreWithComputed();
    const storeOnlyObservables = new StoreOnlyObservables();
    const storeWithoutLog = new StoreWithoutLog();

    let isDividedBy3 = false;
    let isEven = false;
    autorun(() => {
      isDividedBy3 = storeWithComputed.isDividedBy3;
      isEven = storeWithoutLog.isEven;
    });

    storeWithoutLog.increment();
    storeWithComputed.increment();
    storeOnlyObservables.init();
    expect(isDividedBy3).toBeFalsy();
    expect(isEven).toBeTruthy();
    expect(storeWithComputed.value).toBe(1);
    expect(collectingWriter.history).toMatchSnapshot();

    storeWithComputed.increment();
    storeWithComputed.increment();

    expect(collectingWriter.history).toMatchSnapshot();
  });

  it('logs array changes correctly', () => {
    const todoStore = new TodoStore();
    todoStore.addTodo(1, 'Play balalaika');

    expect(collectingWriter.history).toMatchSnapshot();

    todoStore.markAsDone(1);

    expect(collectingWriter.history).toMatchSnapshot();
  });

  it('logs plain array changes correctly', () => {
    const storeOnlyObservables = new StoreOnlyObservables();
    storeOnlyObservables.add();
    storeOnlyObservables.add();
    storeOnlyObservables.removeFirst();
    storeOnlyObservables.removeLast();
    storeOnlyObservables.add();
    storeOnlyObservables.addAndRemoveSameAction();
    storeOnlyObservables.spliceFirst();
    storeOnlyObservables.setMultiple();

    expect(collectingWriter.history).toMatchSnapshot();
  });

  it('logs observable and observable map', () => {
    const store = new ParticipantStore();
    store.onlineUserSet(1, true);
    store.onlineUserSet(2, true);
    store.onlineUserDelete(1);
    store.onlineUserSet(2, false);

    expect(collectingWriter.history).toMatchSnapshot();
  });

  it('logs observable set', () => {
    const store = new ParticipantStore();
    store.adminUserAdd(1);
    store.adminUserAdd(2);
    store.adminUserAdd(2);
    store.adminUserAdd(3);
    store.adminUserDelete(2);
    store.adminUserDelete(3);

    expect(collectingWriter.history).toMatchSnapshot();
  });

  it('works with factory function store', () => {
    const store = createCounterStore();
    store.increment();
    store.increment();
    store.increment();

    expect(collectingWriter.history).toMatchSnapshot();
  });

  it('factory function store with no loggable name throws', () => {
    const createCounterWithoutLoggableName = () => {
      return makeLoggable(
        makeAutoObservable({
          count: 0,
          increment() {
            this.count++;
          },
        })
      );
    };

    expect(createCounterWithoutLoggableName).not.toThrow(Error);
  });

  it('non observable factory function store throws', () => {
    const createCounterNotObservable = () => {
      return makeLoggable({
        count: 0,
        increment() {
          this.count++;
        },
      });
    };

    expect(createCounterNotObservable).toThrow(Error);
  });

  it('does not log action with class properties', () => {
    class StoreDestructuringClassProperties {
      value = 0;

      constructor() {
        makeAutoObservable(this);
        makeLoggable(this);
      }

      increment = () => {
        this.value++;
      }
    }

    const store = new StoreDestructuringClassProperties();
    const { increment } = store;

    increment();
    increment();

    expect(collectingWriter.history).toMatchSnapshot();
  })

  it('logs action with { autoBind: true }', () => {
    class StoreDestructuringAutoBind {
      value = 0;

      constructor() {
        makeAutoObservable(this, {}, { autoBind: true });
        makeLoggable(this);
      }

      increment() {
        this.value++;
      }
    }

    const store = new StoreDestructuringAutoBind();
    const { increment } = store;

    increment();
    increment();

    expect(collectingWriter.history).toMatchSnapshot();
  })
});
