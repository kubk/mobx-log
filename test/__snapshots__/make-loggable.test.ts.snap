// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeLoggable does not log action with class properties 1`] = `
Array [
  "%c[O] color:#FFBC2D StoreDestructuringClassProperties.value 0 -> 1",
  "%c[O] color:#FFBC2D StoreDestructuringClassProperties.value 1 -> 2",
]
`;

exports[`makeLoggable logs 1`] = `
Array [
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 [object Object] -> true",
  "%c[A]%c StoreWithComputed.increment() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreWithComputed.value 0 -> 1",
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 true -> false",
  "%c[A]%c StoreOnlyObservables.init() color:#FF6157 color:inherit",
]
`;

exports[`makeLoggable logs 2`] = `
Array [
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 [object Object] -> true",
  "%c[A]%c StoreWithComputed.increment() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreWithComputed.value 0 -> 1",
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 true -> false",
  "%c[A]%c StoreOnlyObservables.init() color:#FF6157 color:inherit",
  "%c[A]%c StoreWithComputed.increment() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreWithComputed.value 1 -> 2",
  "%c[A]%c StoreWithComputed.increment() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreWithComputed.value 2 -> 3",
  "%c[C] color:#28B7F6 StoreWithComputed.isDividedBy3 false -> true",
]
`;

exports[`makeLoggable logs action with { autoBind: true } 1`] = `
Array [
  "%c[A]%c StoreDestructuringAutoBind.increment() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreDestructuringAutoBind.value 0 -> 1",
  "%c[A]%c StoreDestructuringAutoBind.increment() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreDestructuringAutoBind.value 1 -> 2",
]
`;

exports[`makeLoggable logs array changes correctly 1`] = `
Array [
  "%c[A]%c TodoStore.addTodo(%o, %o) color:#FF6157 color:inherit 1 Play balalaika",
  "%c[O] color:#FFBC2D TodoStore.todos has changed. Added:  [object Object]",
]
`;

exports[`makeLoggable logs array changes correctly 2`] = `
Array [
  "%c[A]%c TodoStore.addTodo(%o, %o) color:#FF6157 color:inherit 1 Play balalaika",
  "%c[O] color:#FFBC2D TodoStore.todos has changed. Added:  [object Object]",
  "%c[A]%c TodoStore.markAsDone(%o) color:#FF6157 color:inherit 1",
  "%c[O] color:#FFBC2D TodoStore.todos[..].isDone false -> true",
]
`;

exports[`makeLoggable logs observable and observable map 1`] = `
Array [
  "%c[A]%c ParticipantStore.onlineUserSet(%o, %o) color:#FF6157 color:inherit 1 true",
  "%c[O] color:#FFBC2D ParticipantStore.onlineUsers - added 1 : true",
  "%c[A]%c ParticipantStore.onlineUserSet(%o, %o) color:#FF6157 color:inherit 2 true",
  "%c[O] color:#FFBC2D ParticipantStore.onlineUsers - added 2 : true",
  "%c[A]%c ParticipantStore.onlineUserDelete(%o) color:#FF6157 color:inherit 1",
  "%c[O] color:#FFBC2D ParticipantStore.onlineUsers - deleted 1",
  "%c[A]%c ParticipantStore.onlineUserSet(%o, %o) color:#FF6157 color:inherit 2 false",
  "%c[O] color:#FFBC2D ParticipantStore.onlineUsers - updated 2 value true -> false",
]
`;

exports[`makeLoggable logs observable set 1`] = `
Array [
  "%c[A]%c ParticipantStore.adminUserAdd(%o) color:#FF6157 color:inherit 1",
  "%c[O] color:#FFBC2D ParticipantStore.adminUsers - added 1",
  "%c[A]%c ParticipantStore.adminUserAdd(%o) color:#FF6157 color:inherit 2",
  "%c[O] color:#FFBC2D ParticipantStore.adminUsers - added 2",
  "%c[A]%c ParticipantStore.adminUserAdd(%o) color:#FF6157 color:inherit 2",
  "%c[A]%c ParticipantStore.adminUserAdd(%o) color:#FF6157 color:inherit 3",
  "%c[O] color:#FFBC2D ParticipantStore.adminUsers - added 3",
  "%c[A]%c ParticipantStore.adminUserDelete(%o) color:#FF6157 color:inherit 2",
  "%c[O] color:#FFBC2D ParticipantStore.adminUsers - removed 2",
  "%c[A]%c ParticipantStore.adminUserDelete(%o) color:#FF6157 color:inherit 3",
  "%c[O] color:#FFBC2D ParticipantStore.adminUsers - removed 3",
]
`;

exports[`makeLoggable logs plain array changes correctly 1`] = `
Array [
  "%c[A]%c StoreOnlyObservables.add() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Added:  1",
  "%c[A]%c StoreOnlyObservables.add() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Added:  2",
  "%c[A]%c StoreOnlyObservables.removeFirst() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Removed:  1",
  "%c[A]%c StoreOnlyObservables.removeLast() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Removed:  2",
  "%c[A]%c StoreOnlyObservables.add() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Added:  3",
  "%c[A]%c StoreOnlyObservables.addAndRemoveSameAction() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Added:  4",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Removed:  3",
  "%c[A]%c StoreOnlyObservables.spliceFirst() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Removed:  4",
  "%c[A]%c StoreOnlyObservables.setMultiple() color:#FF6157 color:inherit",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers has changed. Added:  1,2,3,4,5,6",
  "%c[O] color:#FFBC2D StoreOnlyObservables.numbers  -> 1,2,3,4,5,6",
]
`;

exports[`makeLoggable logs store with destructuring 1`] = `
Array [
  "%c[O] color:#FFBC2D StoreDestructuring.value 0 -> 1",
  "%c[O] color:#FFBC2D StoreDestructuring.value 1 -> 2",
]
`;

exports[`makeLoggable works with factory function store 1`] = `
Array [
  "%c[A]%c counter.increment() color:#FF6157 color:inherit",
  "%c[A]%c counter.increment() color:#FF6157 color:inherit",
  "%c[A]%c counter.increment() color:#FF6157 color:inherit",
]
`;
